### PR TITLE
tac: read files in reverse in chunks

### LINKED ISSUE #2262: tac: output differs from GNU for -b option
Environment: Ubuntu 18.04, GNU coreutils v8.28, `uutils` commit 726f2712732bb086db1dca21ffa01ab1635cb791.

The `tac -b` command currently disagrees with the GNU version in some cases.

Failing unit tests:
```rust
#[test]
fn test_before_trailing_separator_no_leading_separator() {
    new_ucmd!().arg("-b").pipe_in("a\nb\n").succeeds().stdout_is("\n\nba");  // currently outputs" \nb\na"
}

#[test]
fn test_before_trailing_separator_and_leading_separator() {
    new_ucmd!().arg("-b").pipe_in("\na\nb\n").succeeds().stdout_is("\n\nb\na");  // currently outputs "\nb\na\n"
}

#[test]
fn test_before_leading_separator_no_trailing_separator() {
    new_ucmd!().arg("-b").pipe_in("\na\nb").succeeds().stdout_is("\nb\na");  // currently output "b\na\n"
}

#[test]
fn test_before_no_separator() {
    new_ucmd!().arg("-b").pipe_in("ab").succeeds().stdout_is("ab");  // currently outputs "\nab"
}
```

### PR DESCRIPTION
**Pull request by jfinkels**
This pull request changes the implementation of `tac` when reading from a file: read the file in chunks, starting from the end of the file, and stitching together lines as they are discovered.

This change also fixes a bug in the behavior of `tac -b`. Previously, it was not assuming that the line separators were at the *beginning* of each line, which caused its output to disagree with GNU `tac`. Fixes #2262.

Finally, this pull request includes a commit that refactors the `rchunks()` function and necessary structs into `uucore`. This replaces code shared by both the `tac` and the `tail` programs, eliminating duplicate code.

### CODE REVIEW

File: src/uu/tac/src/tac.rs [hunk 1]
```diff
@@ -82,96 +86,62 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
...
+                        show_error!("failed to open '{}' for reading: {}", filename, e)
+                    }
+                    _ => show_error!("failed to read '{}': {}", filename, e),
                 }
+                exit_code = 1;
             }
-        });
-
-        let mut data = Vec::new();
-        if let Err(e) = file.read_to_end(&mut data) {
```
Comment (jfinkels) [inline src/uu/tac/src/tac.rs:119]: Previously, the code was reading the entire file into memory. This pull request replaces that with reading a file by chunks in reverse. (But for `stdin` we still read the entire stream into memory.)

File: tests/by-util/test_tac.rs [hunk 2]
```diff
@@ -23,7 +23,7 @@ fn test_stdin_non_newline_separator_before() {
         .args(&["-b", "-s", ":"])
         .pipe_in("100:200:300:400:500")
         .run()
-        .stdout_is("500:400:300:200:100");
+        .stdout_is(":500:400:300:200100");
```
Comment (jfinkels) [inline tests/by-util/test_tac.rs:26]: I changed this test to match the output of `GNU` tac:
```sh
$ printf "100:200:300:400:500" | tac -b -s ":"
:500:400:300:200100
```

File: tests/fixtures/tac/delimited_primes_before.expected [hunk 3]
```diff
@@ -1 +1 @@
-97:89:83:79:73:71:67:61:59:53:47:43:41:37:31:29:23:19:17:13:11:7:5:3:2
\ No newline at end of file
+:97:89:83:79:73:71:67:61:59:53:47:43:41:37:31:29:23:19:17:13:11:7:5:32
```
Comment (jfinkels) [inline tests/fixtures/tac/delimited_primes_before.expected:1]: same correction had to be made here to match GNU:
```
$ printf "2:3:5:7:11:13:17:19:23:29:31:37:41:43:47:53:59:61:67:71:73:79:83:89:97" | tac -b -s ":"
:97:89:83:79:73:71:67:61:59:53:47:43:41:37:31:29:23:19:17:13:11:7:5:32
```

**Comment by sylvestre** [comment]
Sorry, you are doing this for perf reason?

btw, it fails:

```
---- test_tac::test_invalid_input stdout ----
current_directory_resolved: 
run: D:\a\coreutils\coreutils\target\i686-pc-windows-msvc\debug\coreutils.exe tac b
thread 'test_tac::test_invalid_input' panicked at ''tac: failed to open 'b' for reading: The system cannot find the file specified. (os error 2)
' does not contain 'failed to open 'b' for reading: No such file or directory'', tests\common\util.rs:308:9


```

**Comment by sylvestre** [comment]
@jfinkels ping?

**Comment by jfinkels** [comment]
I'm going to close this and separate this into two separate pull requests: one for correcting the behavior for the `-b` option, and another that implements the performance improvement for reading a file in reverse. For the latter, I will add a report that demonstrates the performance improvement. Sorry for the incomplete pull request here.

**Pull request closed without merging**